core/metacling/src/TCling.cxx

```diff
@@ -6919,14 +6943,87 @@ const char* TCling::GetClassSharedLibs(const char* cls)
+/// Note: In order to speed up the search, we display the dependencies of the
+/// libraries which are not yet loaded. For instance, if libB.so was loaded
+/// the list would contain: lib libA.so libC.so.
+static std::string GetSharedLibDepsSlow(std::string lib,
+                                        cling::Interpreter *interp,
+                                        bool skipLoadedLibs = true)
+{
+   TString LibFullPath(lib);
+   if (!llvm::sys::path::is_absolute(lib)) {
+      if (!gSystem->FindDynamicLibrary(LibFullPath, /*quiet=*/true)) {
+         Error("GetSharedLibDepsForModule", "Cannot find library '%s'", lib.c_str());
+         return "";
+      }
+   } else {
+      lib = llvm::sys::path::filename(lib);
+   }
+
+   auto ObjF = llvm::object::ObjectFile::createObjectFile(LibFullPath.Data());
+   if (!ObjF)
+      return "";
+
+   llvm::object::ObjectFile *BinObjFile = ObjF.get().getBinary();
+
+   std::set<string> DedupSet;
+   std::string Result = lib + ' ';
+   for (const auto &S : BinObjFile->symbols()) {
+      uint32_t Flags = S.getFlags();
+      if (Flags & llvm::object::SymbolRef::SF_Undefined) {
+         llvm::Expected<StringRef> SymNameErr = S.getName();
+         if (!SymNameErr) {
+            Warning("GetSharedLibDepsForModule", "Failed to read symbol");
+            continue;
+         }
+         llvm::StringRef SymName = SymNameErr.get();
+         if (SymName.empty())
+            continue;
+
+         // If we can find the address of the symbol, we have loaded it. Skip.
+         if (skipLoadedLibs && llvm::sys::DynamicLibrary::SearchForAddressOfSymbol(SymName))
+            continue;
+
+         std::string found
+            = SearchSharedLibraries(SymName, interp, /*searchSystem*/false);
+         // The expected output is just filename without the full path, which
+         // is not very accurate, because our Dyld implementation might find
+         // a match in location a/b/c.so and if we return just c.so ROOT might
+         // resolve it to y/z/c.so and there we might not be ABI compatible.
+         // FIXME: Teach the users of GetSharedLibDeps to work with full paths.
+         if (!found.empty()) {
+            std::string cand = llvm::sys::path::filename(found).str();
+            if (!DedupSet.insert(cand).second)
+               continue;
+
+            Result += cand + ' ';
+         }
+      }
+   }
+
+   return Result;
+}
+
+
 ////////////////////////////////////////////////////////////////////////////////
 /// Get the list a libraries on which the specified lib depends. The
 /// returned string contains as first element the lib itself.
 /// Returns 0 in case the lib does not exist or does not have
 /// any dependencies.
 
-const char* TCling::GetSharedLibDeps(const char* lib)
+const char* TCling::GetSharedLibDeps(const char* lib, bool useDyld/* = false*/)
 {
+   if (useDyld) {
+      std::string libs = GetSharedLibDepsSlow(lib, fInterpreter);
+      if (!libs.empty()) {
```

Comment:
This looks more like `tryDyld`: the algorithm falls back to the `!useDyld` case if the symbol lookup fails. Maybe rename the parameter?

Reply:
I've changed it but I am not convinced it conveys better the meaning.